R/preview-standalone.R

```diff
@@ -436,3 +436,83 @@ st_image_show <- function(path,
   )
   return(ans)
 }
+
+#' Render and save a table in png or pdf format
+#'
+#' @param x an stable object.
+#' @param file the final output file name; if `file` has `.tex` extension, the
+#' extension will be replaced with the value in `format`.
+#' @param stem used to form the final output file name; this argument will
+#' override what is provided in `file`.
+#' @param format the output file format.
+#' @param dir the final output directory.
+#' @param build_dir the directory where the image will be built.
+#' @param ... passed to [st_aspdf()] or [st_aspng()], depending on the value of
+#' `format`.
+#'
+#' @details
+#' The table will be built / rendered in `tempdir()`.
+#'
+#' @examples
+#' tab <- stable(stdata(), output_file = "foo.tex")
+#'
+#' \dontrun{
+#'   ans <- stable_save_image(tab format = "pdf")
+#'   basename(ans)
+#'   st_image_show(ans)
+#' }
+#'
+#' @seealso [st_as_image()]
+#'
+#' @export
+stable_save_image <- function(x,
+                              file = attr(x, "stable_file"),
+                              stem = NULL,
+                              format = c("png", "pdf"),
+                              dir = getOption("pmtables.dir"),
+                              build_dir = tempdir(), ...) {
+
+  if(!inherits(x, "stable")) {
+    stop(
+      "x is not an 'stable' object; ",
+      "maybe this object was corrupted or it wasn't generated from 'stable()'",
+      call.=FALSE
+    )
+  }
+  format <- match.arg(format)
+  if(is.character(file)) {
+    ext <- file_ext(file)
+    if(ext=="tex") {
+      stem <- file_path_sans_ext(file)
+      file <- paste0(stem, format)
+    }
+  }
+  if(is.character(stem)) {
+    file <- paste0(stem, ".", format)
+  }
+  if(is.null(file)) {
+    stop("The output file name could not be determined.", call. = FALSE)
+  }
+  if(format=="pdf") {
+    ans <- st_aspdf(
+      x,
+      stem = file_path_sans_ext(file),
+      dir = build_dir,
+      ...
+    )
+  } else {
+    ans <- st_aspng(
+      x,
+      stem = file_path_sans_ext(file),
+      dir = build_dir,
+      ...
+    )
+  }
+  if(!is.null(dir)) {
+    file <- file.path(dir, file)
```

Comment:
Based on the logic on this line and light testing, I think the expectation is that `file` is just a base name and never contains any directory components. Should this function guard against that and give a more informative error?

```
stable_save_image(stable(stdata()), file = "/tmp/foo.pdf")
#> Warning messages:
#> 1: In system2(command = command, args = args, stdout = TRUE, stderr = TRUE) :
#>   running command ''latex' -halt-on-error -jobname=/tmp/foo standalone-preview.tex 2>&1' had status 1
#> 2: In st_to_standalone(x, stem, dir, command = "latex", font = font,  :
#>   the standalone preview build didn't work.
#> 3: In file.create(to[okay]) :
#>   cannot create file '/tmp/Rtmpvtnp6Q//tmp/foo.png', reason 'No such file or directory'
```